stanford_soe_r25.module

```diff
@@ -0,0 +1,889 @@
+      if (preg_match('(\d{4}-\d{2}-\d{2} - \d{4}-\d{2}-\d{2})', $blackout) === 1) {
+        $tmp3 = explode(" - ", $blackout);
+        $blackouts[] = array('start' => $tmp3[0], 'end' => $tmp3[1]);
+      }
+    }
+  }
+  return $blackouts;
+}
+
+// function to check if a given date is blacked out based on blackout dates
+// and current date.
+// $date is given as a UNIX timestamp
+function _stanford_soe_r25_date_blacked_out($date) {
+
+  // if an empty date is given, return false
+  if (empty($date)) {
+    return FALSE;
+  }
+
+  // if I'm currently in a blackout period, the requested date doesn't matter - we consider it blacked out.
+  // if I'm *not* currently in a blackout period, then the requested date has to be before the next blackout starts.
+  // if I'm currently past the last possible blackout, consider me blacked out so an admin will update the dates.
+  $blackouts = _stanford_soe_r25_parse_blackout_dates(variable_get('stanford_soe_r25_blackout_dates', ''));
+  $blacked_out = TRUE;
+
+  // first find out if we are currently in a blackout
+  // if not, find out when the next blackout starts
+  $cur = time();
+  $cur_blackout = TRUE;
+  $next_blackout = 0;
+  foreach ($blackouts as $blackout) {
+    if ($cur < strtotime($blackout['start'])) {
+      $cur_blackout = FALSE;
+      $next_blackout = strtotime($blackout['start']);
+      break;
+    }
+    else {
+      if ($cur <= strtotime($blackout['end'])) {
+        break;
+      }
+    }
+  }
+
+  // now see if the requested date is before the next blackout
+  if (!$cur_blackout && $date < $next_blackout) {
+    $blacked_out = FALSE;
+  }
+  return $blacked_out;
+}
+
+// this is a simple function to output room configurations for dev purposes.
+function _stanford_soe_r25_list_rooms() {
+  $rooms = _stanford_soe_r25_room_config_load();
+  $room_output = print_r($rooms, TRUE);
+  $output = 'rooms: <pre>' . $room_output . '</pre>';
+  return $output;
+}
+
+// check if the current user can confirm or cancel a reservation
+function _stanford_soe_r25_user_can_cancel_or_confirm($room_id, $event_id, $op) {
+  // first get the event's XML from 25Live
+  $result = _stanford_soe_r25_api_call('event-get', NULL, NULL, $event_id);
+  if ($result) {
+    // make sure the event is for the requested room so nobody pulls a fast one.
+    $rooms = _stanford_soe_r25_room_config_load();
+    if (empty($result['index']['R25:SPACE_ID']) || !is_array($result['index']['R25:SPACE_ID']) ||
+      $result['vals'][$result['index']['R25:SPACE_ID'][0]]['value'] != $rooms[$room_id]['space_id']
+    ) {
+      drupal_set_message('Room mismatch for confirm or cancel event.', 'error', FALSE);
+      return FALSE;
+    }
+
+    // default is that user can't do operation (confirm or cancel)
+    $can_cancel = FALSE;
+    global $user;
+
+    // if the user is user1 or has administer rights in Drupal, let them do operation
+    if ($user->uid == 1 || user_access('administer stanford r25')) {
```

Comment:
dont need to check for uid 1. user 1 gets all permissions. see https://api.drupal.org/api/drupal/modules!user!user.module/function/user_access/7.x